flow/shared/authenticator.go

```diff
@@ -0,0 +1,218 @@
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/metadata"
+)
+
+type AuthenticationConfig struct {
+	DevKeySetJSON string `json:"key_set_json" yaml:"key_set_json" mapstructure:"key_set_json"`
+	Auth0Domain   string `json:"auth0_domain" yaml:"auth0_domain" mapstructure:"auth0_domain"`
+	Auth0ClientID string `json:"auth0_client_id" yaml:"auth0_client_id" mapstructure:"auth0_client_id"`
+	Enabled       bool   `json:"enabled" yaml:"enabled" mapstructure:"enabled"`
+}
+
+type identityProvider struct {
+	keySet      jwk.Set
+	validateOpt jwt.ValidateOption
+	issuer      string
+}
+
+func AuthGrpcMiddleware() ([]grpc.ServerOption, error) {
+	devKeySetJSON, _ := os.LookupEnv("PEERDB_OAUTH_KEY_SET_JSON")
+	auth0Domain, _ := os.LookupEnv("PEERDB_OAUTH_AUTH0DOMAIN")
+	auth0ClientID, _ := os.LookupEnv("PEERDB_OAUTH_AUTH0CLIENTID")
+	cfg := AuthenticationConfig{
+		Enabled:       devKeySetJSON != "" || auth0Domain != "",
+		DevKeySetJSON: devKeySetJSON,
+		Auth0Domain:   auth0Domain,
+		Auth0ClientID: auth0ClientID,
+	}
+	// load identity providers before checking if authentication is enabled so configuration can be validated
+	ip, err := identityProvidersFromConfig(cfg)
+
+	if !cfg.Enabled {
+		if err != nil { // if there was an error loading identity providers, warn only if authentication is disabled
+			slog.Warn("failed to initialize JWK key set", slog.Any("error", err))
+		}
+
+		slog.Warn("authentication is disabled")
+
+		return nil, nil
+	}
+
+	if err != nil {
+		return nil, err
+	}
+
+	return []grpc.ServerOption{
+		grpc.ChainUnaryInterceptor(func(ctx context.Context, req any, info *grpc.UnaryServerInfo, handler grpc.UnaryHandler) (any, error) {
+			var authHeader string
+			authHeaders := metadata.ValueFromIncomingContext(ctx, "authorization")
+			if len(authHeaders) == 1 {
+				authHeader = authHeaders[0]
+			} else if len(authHeaders) > 1 {
+				return nil, errors.New("multiple Authorization headers supplied, request rejected")
+			}
+			_, err := validateRequestToken(authHeader, ip...)
+			if err != nil {
+				slog.Debug("failed to validate request token", slog.Any("error", err))
+				return nil, err
+			}
+			return handler(ctx, req)
+		}),
+	}, nil
+}
+
+func validateRequestToken(authHeader string, ip ...identityProvider) ([]byte, error) {
+	payload, err := jwtFromRequest(authHeader)
+	if err != nil {
+		return nil, fmt.Errorf("failed to parse authorization header: %w", err)
+	}
+
+	token, err := jwt.ParseInsecure(payload)
+	if err != nil {
+		return nil, fmt.Errorf("failed to parse token: %w", err)
+	}
+
+	provider, err := identityProviderByToken(ip, token)
+	if err != nil {
+		return nil, err
+	}
+
+	validateOpts := identityProviderValidateOpts(provider)
+	if err := jwt.Validate(token, validateOpts...); err != nil {
+		return nil, fmt.Errorf("failed to validate token: %w", err)
+	}
+
+	if _, err := jws.Verify(payload, jws.WithKeySet(provider.keySet)); err != nil {
+		return nil, fmt.Errorf("failed to verify token: %w", err)
+	}
+
```

Comment:
TODO: We need to add additionally validation against some custom claims.
We could take something like `..._DESIRED_CLAIM_KEY` and `..._DESIRED_CLAIM_VALUE` and match it against the jwt

Reply:
why isn't this upstream?

Reply:
no clue, but I think we do need it as per the discussion I had

Reply:
My opinion is that it doesn't go far enough at the moment. I'm not sure why they didn't implement upstream. It would make me feel more comfortable if there was some data section checking here, but we can wait to see what is in that payload.